### PR TITLE
179/implementation of missing latitude and longitude at crawling

### PR DESCRIPTION
**Pull request by AndreBruecke**
Created enhancer for missing geodata.

Enhancement script is being called as part of the default pipeline step for every scraper file.

### CODE REVIEW

**Review by jonas-hda — CHANGES_REQUESTED** [review]
Please change the PR in a way that it only contains files related to the feature. This PR is for the latitude/longitude enhancement, but contains implementation for the weltwaerts crawler, github template, base scraper... . I assume this is due to creating the branch from master and not staging? alternatively I can work the lat long enhancement changes directly into branch 176/Design_proposal_of_changes_to_the-system_architecture_for_the_enhancement_process_adress_to_process_data_from_multiple_crawlers and we abort the PR afterwards

**Comment by FiedlerJan** [comment]
> 
> 
> Please change the PR in a way that it only contains files related to the feature. This PR is for the latitude/longitude enhancement, but contains implementation for the weltwaerts crawler, github template, base scraper... . I assume this is due to creating the branch from master and not staging? alternatively I can work the lat long enhancement changes directly into branch 176/Design_proposal_of_changes_to_the-system_architecture_for_the_enhancement_process_adress_to_process_data_from_multiple_crawlers and we abort the PR afterwards

Just approve #192 and this problem dones not exist.

**Comment by jonas-hda** [comment]
> 
> 
> > Please change the PR in a way that it only contains files related to the feature. This PR is for the latitude/longitude enhancement, but contains implementation for the weltwaerts crawler, github template, base scraper... . I assume this is due to creating the branch from master and not staging? alternatively I can work the lat long enhancement changes directly into branch 176/Design_proposal_of_changes_to_the-system_architecture_for_the_enhancement_process_adress_to_process_data_from_multiple_crawlers and we abort the PR afterwards
> 
> Just approve #192 and this problem dones not exist.



> 
> 
> > Please change the PR in a way that it only contains files related to the feature. This PR is for the latitude/longitude enhancement, but contains implementation for the weltwaerts crawler, github template, base scraper... . I assume this is due to creating the branch from master and not staging? alternatively I can work the lat long enhancement changes directly into branch 176/Design_proposal_of_changes_to_the-system_architecture_for_the_enhancement_process_adress_to_process_data_from_multiple_crawlers and we abort the PR afterwards
> 
> Just approve #192 and this problem dones not exist.

Merge of #192 into staging did not resolve this

**Comment by FiedlerJan** [comment]
> Merge of #192 into staging did not resolve this

Now it's resolved.

File: etl/data_enhancement/domain/default.py [hunk 1]
```diff
@@ -68,3 +70,13 @@ def rank_tags(file):
...
+
+
+# Enhance the geo_location
+def add_lat_lon(file):
+    enhancer = LatLonEnhancer()
+
+    index = 1
+    for post in file:
+        post = enhancer.enhance(post)
+        index += 1
```
Comment (FiedlerJan) [inline etl/data_enhancement/domain/default.py:82]: I guess the index can be removed.

File: etl/data_enhancement/utils/lat_lon_enhancer.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
+from geopy.geocoders import Nominatim
+import csv
+import time
+
+
+class LatLonEnhancer:
```
Comment (FiedlerJan) [inline etl/data_enhancement/utils/lat_lon_enhancer.py:6]: Missing docstring.

File: etl/data_enhancement/utils/lat_lon_enhancer.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    dict_file = 'data_enhancement/output/geocoder_lat_lon.csv'
+
+    lat_lon_dict = {}
+
+    def __init__(self):
```
Comment (FiedlerJan) [inline etl/data_enhancement/utils/lat_lon_enhancer.py:11]: Missing docstring.

File: etl/data_enhancement/utils/lat_lon_enhancer.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        self.geo_locator = Nominatim(user_agent="einander-helfen.org")
+        self.__load_local_storage()
+
+    def enhance(self, post):
```
Comment (FiedlerJan) [inline etl/data_enhancement/utils/lat_lon_enhancer.py:15]: Missing docstring.

File: etl/data_enhancement/utils/lat_lon_enhancer.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+            location = self.geo_locator.geocode(request_string)
+
+            geo_location = {'lat': location.latitude, 'lon': location.longitude}
+            time.sleep(1)
+
+            return geo_location
+        return None
+
+    @staticmethod
+    def get_api_request_string(post):
```
Comment (FiedlerJan) [inline etl/data_enhancement/utils/lat_lon_enhancer.py:80]: Why is this method static?

**Review by FiedlerJan — CHANGES_REQUESTED** [review]
Please add geopy to the requirements.txt

File: etl/data_enhancement/utils/lat_lon_enhancer.py (hunk 5 above)
Comment (AndreBruecke) [reply etl/data_enhancement/utils/lat_lon_enhancer.py:85]: Using a static method was suggested by PyCharm IDE in this case. Could be changed back to private method though.

Comment (jonas-hda) [reply etl/data_enhancement/utils/lat_lon_enhancer.py:85]: static methods are slightly more efficient which is why PyCharm is suggesting it. the benefits are very negligible in practice though.

**Review by FiedlerJan — APPROVED** [review]

**Review by jonas-hda — APPROVED** [review]

**Pull request merged**